### ISSUE TITLE
Empty $config->httpHosts is not allowed for https sites

### ISSUE THREAD

**Issue by ivangretsky** [processwire/processwire-issues#358]
### Short description of the issue

If I set `$config->httpHosts = array()` on an https site the **Unrecognized HTTP host** error message never goes away. If I visit the same site via http, there is no error message.

### Setup/Environment

* ProcessWire version: 3.0.73

**Comment by ryancramerdesign** [comment]
If you've got an empty `$config->httpHosts` array, make sure that you are setting the `$config->httpHost` (string), and that it's consistent with what PHP can see for either `$_SERVER['HTTP_HOST']` or `$_SERVER['SERVER_NAME']`. I'm guessing that in your case either `$config->httpHost` is not set, or that PHP can't see those server vars when https is active. If you want to have PW automatically populate `$config->httpHost` then you'd want populate your `$config->httpHosts` array with all the possible hostnames for your site. In fact, I think that's a better way to go rather than setting the httpHost, unless the site will only ever be accessed at a single host name.

**Comment by ivangretsky** [comment]
Sorry for not keeping up the discussion @ryancramerdesign. Did not have opportunity to work on this one before today.

I did check all the variables you mentioned. Just echoed them in the template file:
$_SERVER["HTTP_HOST"] = examplesite.ru
$_SERVER["SERVER_NAME"] = examplesite.ru
$config->httpHost = examplesite.ru:443

Should that port number be there?

I do not want to include all the hosts in $config->httpHosts, as the site will be accessed from different subdomains. The list will be expandable, so changing something in the config is not desirable. Everything works as it should from http, why not from https?

**Comment by netcarver** [comment]
@ryancramerdesign Not sure if you spotted the question in Ivan's reply.

@ivangretsky Is this still an issue for you with newer versions of PW?

**Comment by ivangretsky** [comment]
This is still an issue. I would prefer not to define `$config->httpHosts` for https sites same way I do it for http sites. I think that should be an option, not a requirement. It makes development and deployment a little harder.

**Comment by ryancramerdesign** [comment]
@ivangretsky @netcarver It might be that the extra 443 appearing in there is the source of the issue. I see why it's being added so am committing an update that prevents that. However, letting a site run with an undefined HTTP host means that the client side gets to define what the HTTP host is. That comes through as a request header in user input. So it's not particularly safe to let things run that way. What I'd suggest is adding some logic in your site/config.php file that sets the `$config->httpHost` manually if it matches an expected regular expression pattern. Something like this:
~~~~~
if(!isset($_SERVER['HTTP_HOST'])) {
  // likely running in CLI mode
  $config->httpHost = 'examplesite.ru';
} else if(preg_match('/^[a-z]{3,20}\.examplesite\.ru$/i', $_SERVER['HTTP_HOST'])) {
  // host name matches abc.examplesite.ru where abc is between 3 to 20 a-z characters
  $config->httpHost = strtolower($_SERVER['HTTP_HOST']);
} else {
  // fallback if host does not match expected regex
  $config->httpHost = 'examplesite.ru';
}
~~~~~
If you don't understand the regex above, let me know more specifically what your range of options is for the host name and I'm happy to adjust the regex in the example above. 

Also, note PW doesn't need the `$config->httpHosts` array if you manually define the `$config->httpHost` like above.

**Comment by netcarver** [comment]
@ivangretsky Please could you review Ryan's change and test it out. If this fixes this issue for you, please go ahead and close it or further the conversation if not.

/remind me to review this issue in 1 week

**Comment by netcarver** [comment]
Closing. @ivangretsky please post more if you need to.

**Comment by ivangretsky** [comment]
The issue seems to be fixed. Tried it on a fresh site without $config->httpHosts defined and no error in admin. Thank you!

Thanks for you help with issues, @netcarver. But have to say, you're a bit harsh pushing to close them. The fixes were in the new release (I am not used to install ProcessWire from certain commits). So I only got 3 working days to check.  Maybe loosen it up a bit and set reminders in a week or two?
